### ISSUE TITLE
Tutorial 6: Attention is all you need

### ISSUE THREAD

**Issue by Hannibal046** [bentrevett/pytorch-seq2seq#170]
Hello, is there any good hyper parameters for transformer with encoder layers and decoder layers of 6? 
By simply change from  
```python
ENC_LAYERS = 3
DEC_LAYERS = 3
N_EPOCHS = 10
```
to 
```python
ENC_LAYERS = 6
DEC_LAYERS = 6
N_EPOCHS = 50
```
I got the following results:

<img width="428" alt="image" src="https://user-images.githubusercontent.com/38466901/127809455-deb9c685-7336-41ad-a8d9-2b363024879c.png">
<img width="613" alt="image" src="https://user-images.githubusercontent.com/38466901/127809595-f1fe5e98-e1e6-48c4-8dcd-978033082670.png">

Do you guys have any idea to get better results with bigger model? Thanks so much !

**Comment by bentrevett** [comment]
Training large Transformer models still requires quite a few tricks to make them work properly. 

The main two techniques I'd recommend looking into are initialization and learning rate schedulers.

A common way to initialize Transformers, which I'd recommend, is using a truncated normal distribution with a std of 0.02.

For learning rate schedulers, a lot of people use a "warm-up" stage, where the learning rate linearly increased from zero to some maximum value, followed by an annealing stage where the learning rate is then slowly decreased (usually linearly too).

**Comment by Yuvaraj91** [comment]
could you share a link or basic code on how we can implement the warm-up stage please?

**Comment by Hannibal046** [comment]
perhaps you can do it like this:
```python
def get_inverse_sqrt_schedule_with_warmup(optimizer, num_warmup_steps, num_training_steps, last_epoch=-1):
    """ Create a schedule with a learning rate that linearly increases from some initial learning rate (``--warmup-init-lr``) until the configured
    learning rate (``--lr``). Thereafter we decay proportional to the inverse square root of
    the number of updates.
    """

    def lr_lambda(current_step):
        current_step = max(1, current_step)
        return min(float(current_step)**-0.5, float(current_step)*(float(num_warmup_steps)**-1.5))
    
    return LambdaLR(optimizer, lr_lambda, last_epoch)
```
For further detail, you can check out this repo https://github.com/jcyk/copyisallyouneed
